File: go-echo/prover.go
package main

import (
	"bytes"
	"errors"
	"os"
	"os/exec"
	"strconv"
	"strings"
)

type Result struct {
	Msg string
	Img []byte
	Tex string
}

func prove(sequent, memory string, timeout int, enableNotification bool) (*Result, error) {
	// lock
	mutex.Lock()
	defer mutex.Unlock()

	// create temp dir
	dir, err := os.MkdirTemp(".", "")
	if err != nil {
		return nil, err
	}
	defer os.RemoveAll(dir)

	// change dir
	if err := os.Chdir(dir); err != nil {
		return nil, err
	}
	defer os.Chdir("..")

	// symlink ../prover
	if err := os.Symlink("../prover", "prover"); err != nil {
		return nil, err
	}

	// run prover
	msg, err := runProver(sequent, memory, timeout)
	if err != nil {
		return nil, err
	}
	// make dvi
	msgDVI, err := makeDVI()
	if err != nil {
		return nil, err
	}
	// make png
	msgPNG, err := makePNG()
	if err != nil {
		return nil, err
	}
	msg += msgDVI + msgPNG

	// create response
	res := &Result{
		Msg: msg,
	}

	// if out.png exists
	if _, err := os.Stat("out.png"); err == nil {
		if enableNotification {
			notifyLineWithImage(msg)
		}
		// read out.png
		img, err := os.ReadFile("out.png")
		if err != nil {
			return nil, err
		}
		res.Img = img
		// read out.tex
		tex, err := os.ReadFile("out.tex")
		if err != nil {
			return nil, err
		}
		res.Tex = string(tex)
	} else {
		if enableNotification {
			notifyLine(msg)
		}
	}

	return res, nil
}

func runProver(sequent, memory string, timeout int) (string, error) {
	// run the command
	stdout, stderr, err := runCommand("../prover.sh", sequent, memory, strconv.Itoa(timeout))

	// Timeout
	if strings.Contains(stderr, "CPU time limit exceeded") {
		if stdout == "" {
			return "Proof Failed: Timeout.", nil
		} else {
			return stdout + " The proof tree is too large to output: Timeout.", nil
		}
	}
	// OutOfMemoryError
	if strings.Contains(stderr, "OutOfMemoryError") {
		if stdout == "" {
			return "Proof Failed: OutOfMemoryError.", nil
		} else {
			return stdout + " The proof tree is too large to output: OutOfMemoryError.", nil
		}
	}
	// StackOverflowError
	if strings.Contains(stderr, "StackOverflowError") {
		return "Proof Failed: StackOverflowError.", nil
	}
	// other err
	if err != nil {
		return "", errors.New("binary execution error.\n" + "stdout: " + stdout + "\n" + "stderr: " + stderr)
	}
	// success
	return stdout, nil
}

func makeDVI() (string, error) {
	// check if out.tex exists
	if _, err := os.Stat("out.tex"); err != nil {
		return "", nil
	}

	// run the command
	stdout, stderr, commandErr := runCommand("latex", "-halt-on-error", "-interaction=nonstopmode", "out.tex")

	// Dimension too large
	if strings.Contains(stdout, "Dimension too large") {
		return " The proof tree is too large to output: Dimension too large.", nil
	}
	// other err
	if _, err := os.Stat("out.dvi"); err != nil || commandErr != nil {
		return "", errors.New("could not compile tex file.\n" + "stdout: " + stdout + "\n" + "stderr: " + stderr)
	}
	// success
	return "", nil
}

func makePNG() (string, error) {
	// check if out.dvi exists
	if _, err := os.Stat("out.dvi"); err != nil {
		return "", nil
	}

	// run the command
	stdout, stderr, commandErr := runCommand("dvipng", "out.dvi", "-o", "out.png")

	// DVI stack overflow
	if strings.Contains(stderr, "DVI stack overflow") {
		return " The proof tree is too large to output: DVI stack overflow.", nil
	}
	// other err
	if _, err := os.Stat("out.png"); err != nil || commandErr != nil {
		return "", errors.New("could not compile dvi file.\n" + "stdout: " + stdout + "\n" + "stderr: " + stderr)
	}
	// success
	return "", nil
}

func runCommand(name string, arg ...string) (string, string, error) {
	cmd := exec.Command(name, arg...)

	// create a buffer to capture stdout and stderr
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	// run the command
	err := cmd.Run()

	return stdout.String(), stderr.String(), err
}

File: python-flask/prover.py
import os
import subprocess
from subprocess import CalledProcessError

from notify import notify_line


def make_proof_tree(sequent: str, memory: str, timeout: int) -> str:
    return prove(sequent, memory, timeout) + make_dvi() + make_img()


def prove(sequent: str, memory: str, timeout: int) -> str:
    try:
        cmd = ['../prover.sh', sequent, memory, str(timeout)]
        result = subprocess.run(cmd, capture_output=True, check=True, text=True)
        # TeXが生成されないとき
        if not os.path.exists('out.tex'):
            raise CalledProcessError(1, cmd, result.stdout, result.stderr)
        # 正常時
        return result.stdout
    except CalledProcessError as e:
        # Timeout
        if 'CPU time limit exceeded' in e.stderr:
            if not e.stdout:
                return 'Proof Failed: Timeout.'
            else:
                return e.stdout + ' The proof tree is too large to output: Timeout.'
        # OutOfMemoryError
        if 'OutOfMemoryError' in e.stderr:
            if not e.stdout:
                return 'Proof Failed: OutOfMemoryError.'
            else:
                return e.stdout + ' The proof tree is too large to output: OutOfMemoryError.'
        # StackOverflowError
        if 'StackOverflowError' in e.stderr:
            return 'Proof Failed: StackOverflowError.'
        # Other error
        notify_line(f'Error: {e}')
        notify_line(f'stdout: {e.stdout}')
        notify_line(f'stderr: {e.stderr}')
        return 'An unexpected error has occurred: Binary Execution Error. The bug report was sent to the developer.'


def make_dvi() -> str:
    if not os.path.exists('out.tex'):
        return ''
    try:
        cmd = ['latex', '-halt-on-error', '-interaction=nonstopmode', 'out.tex']
        result = subprocess.run(cmd, capture_output=True, check=True, text=True)
        # DVIが生成されないとき
        if not os.path.exists('out.dvi'):
            raise CalledProcessError(1, cmd, result.stdout, result.stderr)
        # 正常時
        return ''
    except CalledProcessError as e:
        # Dimension too large
        if 'Dimension too large' in e.stdout:
            return ' The proof tree is too large to output: Dimension too large.'
        # Other error
        notify_line(f'Error: {e}')
        notify_line(f'stdout: {e.stdout}')
        notify_line(f'stderr: {e.stderr}')
        return ' An unexpected error has occurred: Could not compile tex file. ' \
               'The bug report was sent to the developer.'


def make_img() -> str:
    if not os.path.exists('out.dvi'):
        return ''
    try:
        cmd = ['dvipng', 'out.dvi', '-o', 'out.png']
        result = subprocess.run(cmd, capture_output=True, check=True, text=True)
        # PNGが生成されないとき
        if not os.path.exists('out.png'):
            raise CalledProcessError(1, cmd, result.stdout, result.stderr)
        # 正常時
        return ''
    except CalledProcessError as e:
        # DVI stack overflow
        if 'DVI stack overflow' in e.stderr:
            return ' The proof tree is too large to output: DVI stack overflow.'
        # Other error
        notify_line(f'Error: {e}')
        notify_line(f'stdout: {e.stdout}')
        notify_line(f'stderr: {e.stderr}')
        return ' An unexpected error has occurred: Could not compile dvi file. ' \
               'The bug report was sent to the developer.'
